nixos-rebuild: Allow local builds when --target-host is used again

B4dM4n opened the pull request:
###### Description of changes

This is a followup of #148921, to allow local builds when `--target-host` is used again. It also documents the change in behavior, regarding the specialty of the `localhost` value.

By removing the special handling of an empty `buildHost` and non empty `targetHost`, this change also slightly alters the behavior of `nixos-rebuild`.

Originally by specifying `--target-host target --build-host ""`, the now removed special case would transform those arguments to `--target-host target --build-host target`.
Now the empty `--build-host` would result in a local build.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests)) `nixosTests.installer.simple`
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

cc @l0b0 @veprbl @jtojnar

jtojnar commented on nixos/doc/manual/release-notes/rl-2305.section.md:
```diff
@@ -67,6 +67,8 @@ In addition to numerous new and upgraded packages, this release has the followin
 
 - In `mastodon` it is now necessary to specify location of file with `PostgreSQL` database password. In `services.mastodon.database.passwordFile` parameter default value `/var/lib/mastodon/secrets/db-password` has been changed to `null`.
 
+- The `--target-host` and `--build-host` options of `nixos-rebuild` don't treat the `localhost` value special anymore. The empty string must now be used to force a local deployment/build.
```
The empty string is default so there is no need to force anything with this change.

B4dM4n replied:
The empty string is now the only way to reset to local behavior.

I was thinking about cases where `nixos-rebuild` is wrapped by another script (which is what I do), which pre-fills those options and the user might want to restore the default.

The script might contain something like this:

```shell
exec nixos-rebuild --build-host builder --flake myconfigs#host "$@"
```

B4dM4n replied:
How about this text:

> The `--target-host` and `--build-host` options of `nixos-rebuild` don't treat the `localhost` value special anymore. Local operation is now represented by the empty string.

Or should the "empty string" part simply be removed?

jtojnar replied:
Maybe something like:

> Specifying a target host without a build host will not default the latter to the former. Additionally, the `--target-host` and `--build-host` options of `nixos-rebuild` no longer treat the `localhost` value specially – to build on/deploy to local machine, omit the relevant flag.

B4dM4n replied:
> Specifying a target host without a build host will not default the latter to the former

This wasn't the case before #148921, both values defaulted to local, even if only one was specified.
So this was also true before:

> to build on/deploy to local machine, omit the relevant flag.

jtojnar replied:
> > Specifying a target host without a build host will not default the latter to the former
> 
> This wasn't the case before #148921, both values defaulted to local, even if only one was specified.

Oh, you are right. It was true when I created my deployment script but since then https://github.com/NixOS/nixpkgs/commit/3d34fe9f2b55e29721fa0c0c8ad677c0bf064e3e changed the `buildHost` default.

The second sentence should be sufficient then.

B4dM4n replied:
> It was true when I created my deployment script but since then [3d34fe9](https://github.com/NixOS/nixpkgs/commit/3d34fe9f2b55e29721fa0c0c8ad677c0bf064e3e) changed the `buildHost` default.

I see, this was changed before. I must have build my script afterwards, because I can't remember having this problem.

I updated the changelog with your suggested text, thanks.

B4dM4n committed:
nixos-rebuild: Allow local builds when --target-host is used again

This is a followup of #148921, to allow local builds when
`--target-host` is used again. It also documents the change in
behavior, regarding the specialty of the `localhost` value.

By removing the special handling of an empty `buildHost` and non empty
`targetHost`, this change also slightly alters the behavior of
`nixos-rebuild`.

Originally by specifying `--target-host target --build-host ""`, the
now removed special case would transform those arguments to
`--target-host target --build-host target`.
Now the empty `--build-host` would result in a local build.

arjan-s reviewed (approved):
Yes! Thank you! This fixes the problems I reported in https://github.com/NixOS/nixpkgs/pull/148921#issuecomment-1380667128 .

Mic92 commented:
@ofborg eval
